dnsmasq: add fast-dns-retry and use-stale-cache options to config

user8446 opened the pull request:
DNSMasq v2.88 added `--fast-dns-retry` and `--use-stale-cache option` which would be beneficial to some users.

Changelog: https://thekelleys.org.uk/dnsmasq/CHANGELOG

This just adds those options in `dnsmasq.init` with the config options:  `option fast_dns_retry` and `option use_stale_cache`

I have already been using both of the options without issue. I can also update the [wiki page](https://openwrt.org/docs/guide-user/base-system/dhcp).

Signed-off-by: Glenn Fowler <gfowler1(at)outlook.com>

user8446 committed:
Update dnsmasq.init

DNSMasq v2.88 added `--fast-dns-retry` and `--use-stale-cache option` which would be beneficial to some users.

Changelog: https://thekelleys.org.uk/dnsmasq/CHANGELOG

This just adds the config options `option fast_dns_retry` and `option use_stale_cache`

Signed-off-by: Glenn <18019245+user8446@users.noreply.github.com>

schm0 commented:
This has to be append_bool or?

user8446 committed:
Update dnsmasq.init

Add `option neg_ttl`

Signed-off-by: Glenn <18019245+user8446@users.noreply.github.com>

user8446 commented:
Ok `option neg_ttl` added.

I tested `append_bool` instead of `append_parm` and the behavior is if it is enabled then it uses the defaults but you can't change the defaults. Anything other than `1` or `0` turns it off. `append_bool` of course allows user defined data.

Thoughts? I'm thinking the user should be able to define it.

schm0 commented:
Yes, this is expected behavior as I stated above.
I think this needs some trickery...
Because testing for 0 or 1 when parameters are given like "1000,10000" will throw an error message.
I think the best thing would be, to do this with 2 variables, for example:
fast_dns_retry
fast_dns_retry_params
Use config_get for fast_dns_retry
Check if fast_dns_retry >=2 then use append_parm and use the vaules from fast_dns_retry_params
If fast_dns_retry_params are empty use append_bool instead.
Check if fast_dns_retry = 1 then use append_bool
Or something like that?
Maybe I will check this out deeper later... and update this post.

schm0 commented:
Something like this maybe:
`	config_get_bool fast_dns_retry "$cfg" "fast_dns_retry" 0
	config_get fast_dns_retry_params "$cfg" "fast_dns_retry_params"
	if [ $fast_dns_retry -eq 1 ]; then
		if [ -n "$fast_dns_retry_params" ]; then
				append_parm "$cfg" fast_dns_retry_params "--fast-dns-retry"
		else
				append_bool "$cfg" fast_dns_retry "--fast-dns-retry"
		fi
	fi

	config_get_bool use_stale_cache "$cfg" "use_stale_cache" 0
	config_get stale_cache_params "$cfg" "stale_cache_params"
	if [ $use_stale_cache -eq 1 ]; then
		if [ -n "$stale_cache_params" ]; then
				append_parm "$cfg" stale_cache_params "--use-stale-cache"
		else
				append_bool "$cfg" use_stale_cache "--use-stale-cache"
		fi
	fi`

user8446 commented:
Thanks for the code as I need help to do it the way you want anyway as I'm a PowerShell guy.

I tested the stale cache and `1` does enable it. However, any other number which should enable it and be the `max TTL excess in s` does not.

Also you are correct, a `0` is actually enabled with an infinite max TTL.

schm0 commented:
Use:
option use_stale_cache '1'
option stale_cache_params '86400'
For example

user8446 commented:
Ok I see how you did it now. I tested and it is working as expected with no changes. Nice work!

user8446 committed:
Update dnsmasq.init

Signed-off-by: Glenn <18019245+user8446@users.noreply.github.com>

user8446 committed:
Update dnsmasq.init

Signed-off-by: Glenn <18019245+user8446@users.noreply.github.com>

user8446 commented:
Also just added `option no_round_robin` `--no-round-robin` which was also added in v2.88.

`Dnsmasq normally permutes the order of A or AAAA records for the same name on successive queries, for load-balancing. This turns off that behaviour, so that the records are always returned in the order that they are received from upstream.`

schm0 commented:
In your latest commit you readded
append_parm "$cfg" use_stale_cache "--use-stale-cache"
which is not needed.
And maybe move the neg_ttl options to the other ttl options.

user8446 committed:
Update dnsmasq.init

Group `neg_ttl` with other ttl options

Signed-off-by: Glenn <18019245+user8446@users.noreply.github.com>

user8446 commented:
Ok I checked and the last commit was adding `no_round_robin` : https://github.com/openwrt/openwrt/pull/11738/commits/ac3d681b89d56072466d9abe29b8b0525cf54da3

I grouped `neg_ttl` with the other ttl options: https://github.com/openwrt/openwrt/pull/11738/commits/07e91c1f2a7bb4e009174afefd8347dfd699bfa4

user8446 committed:
dnsmasq: add fast-dns-retry and use-stale-cache options to config

Also `neg_ttl` and `no_round_robin` available options

Signed-off-by: Glenn <gfowler1@outlook.com>

user8446 committed:
append_bool "$cfg" no_round_robin "--no-round-robin"

Also `neg_ttl` and `no_round_robin` options.

Sorry for the duplicate, my settings were not right so it would sign off with my first and last name to pass checks.

Signed-off-by: Glenn Fowler <gfowler1@outlook.com>

user8446 committed:
dnsmasq: add fast-dns-retry and use-stale-cache options to config

Also `neg_ttl` and `no_round_robin` options.

Another attempt to pass checks. Sorry for the commit spamming.

Signed-off-by: Glenn Fowler <gfowler1@outlook.com>

schm0 commented:
I think neg_ttl should be quoted like this:
`append_parm "$cfg" "neg_ttl" "--neg-ttl"`

In the code I posted, the part with the -eq 1 comparison...
I'm not sure if those variables should be quoted too.
I usually quote all my variables...
But I checked the init script and there are some inconsistencies there.
Sometimes variables are quoted and sometimes not.
It shouldn't matter that much because it is an integer comparison?

user8446 committed:
dnsmasq: add fast-dns-retry and use-stale-cache options to config

Quoted variables for consistency.

Signed-off-by: Glenn Fowler <gfowler1@outlook.com>

user8446 commented:
Did a little research and it will work either way unless there is whitespace. All use underscores in the script instead of spaces to eliminate that. 
That would make your habit a good one as quoting them all would be safer and eliminate breakage. I went ahead and quoted them as it makes syntax readability easier too and for consistency.

schm0 commented:
Yes it really only does matter when working with strings.

hauke commented:
This pull request has formal issues, please follow: https://openwrt.org/submitting-patches

yulinbanshi commented:
     这是来自鸿霞的假期自动回复邮件。
您好，你的来信已经收到，现无法马上回复您的邮件。我将尽快给您回复。
     谢谢！

user8446 commented:
Thanks, these were fixed after I saw the checks fail such as my lastname not showing up on the signoff.
